src/operations/blas3/gemm_no_local.hpp

```diff
@@ -394,76 +415,444 @@ class Gemm<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize, tile_type,
+                        do_check<check_col>(in_col(work_per_load - 1));
+        cl::sycl::vec<element_t, work_per_load> in_vec{0};
+        if (in_range) {
+          // if in range perform a vectorised load
+          in_vec.template load<address_t::global_space>(0, ptr + j * ld);
+
+        } else {
+          // if not in range perform element-wise load checking boundaries at
+          // each load.
+#pragma unroll
+          for (int l = 0; l < work_per_load; l++) {
+            if (do_check<check_row>(in_row(i * next_element + j)) &&
+                do_check<check_col>(in_col(l))) {
+              reinterpret_cast<element_t *>(&in_vec)[l] = *(ptr + j * ld + l);
+            }
+          }
+        }
+        // Stores the loaded value in the register while untransposing it.
+#pragma unroll
+        for (int k = 0; k < work_per_load; k++) {
+          reg[j + i * work_per_load + k * rows] =
+              reinterpret_cast<element_t *>(&in_vec)[k];
+        }
+      }
+      ptr += next_element * ld;
+    }
+  }
+
+  /*!
+   * @brief Performs a single load from B into private registers, with the
+   * amount of elements loaded determined by work_per_load. This version of the
+   * function is called if trans == false.
+   * @tparam check_row : determines whether to perform bounds checking in the
+   * row direction.
+   * @tparam check_col : determines whether to perform bounds checking in the
+   * column direction.
+   * @tparam work_per_load : the number of elements loaded at one time, this is
+   * also called the packet or vector size.
+   * @tparam trans : true if B's representation is transposed i.e. it is row
+   * major instead of column.
+   * @tparam PointerType: the type of the input matrix.
+   * @tparam RowCheckType: the type of a function used for checking the
+   * boundary in the row direction for blocks of data located at the edge of the
+   * input matrix.
+   * @tparam ColCheckType: the type of a function used for checking the
+   * boundary in the column direction for blocks of data located at the edge of
+   * the input matrix.
+   * @param ptr : the input matrix B.
+   * @param reg : the private register for B
+   * @param row_ofs: How many rows B has been offset by, used in bounds
+   * checking.
+   * @param col_ofs : How many columns B has been offset by, used in bounds
+   * checking.
+   * @param in_row : function which checks the boundary in the row direction.
+   * @param in_col : function which checks the boundary in the col direction.
+   * @param out_of_range: exits the function early if block is out of range.
+   */
+  template <bool check_row, bool check_col, index_t work_per_load, bool trans,
+            typename PointerType, typename RowCheckType, typename ColCheckType>
+  SYCL_BLAS_INLINE typename std::enable_if<!trans>::type load_single_b(
+      PointerType ptr, element_t *reg, const index_t &row_ofs,
+      const index_t &col_ofs, const RowCheckType &in_row,
```

Comment:
Could `in_row` and `in_col` have better names? `is_valid_row` or something?

Reply:
Changed! Thanks for the suggestion.